Please visit the following website to download example projects:

http://galaxy.ustc.edu.cn:30803/APEC/

(1) **project01** contains single cell samples of hematopoietic stem cell differentiation (including HSC, MPP, CMP, GMP, MEP, LMPP, CLP, and pDC cells), from "Buenrostro, J.D. et al. Integrated Single-Cell Analysis Maps the Continuous Regulatory Landscape of Human Hematopoietic Differentiation. Cell 173, 1535-1548 e1516 (2018)".

How to run:

    #### 1. Cell clustering by APEC algorithm. It takes ~10 minutes.

    from APEC import clustering, plot
    clustering.build_accesson('$project03')
    clustering.cluster_byAccesson('$project03', norm='probability')
    plot.plot_tsne('$project03', cell_label='notes')
    plot.correlation('$project03')

    #### 2. Generate pseudo-time trajectory by monocle. It takes ~4 minutes.

    from APEC import generate, plot
    generate.monocle_trajectory('$project03')
    plot.plot_trajectory('$project03', angles=[58,-103])

    #### 3 Plot genes on tSNE. It takes several minutes.

    from APEC import generate, plot
    generate.gene_score('$project03', genome='hg19')
    plot.plot_feature('$project03', space='tsne', feature='gene', name='FOXO1')

    #### 4. Plot motifs on trajectory. It takes ~1 minute for each motif.

    from APEC import plot
    plot.plot_feature('$project03', space='trajectory', feature='motif', name='Hoxa9', angles=[58,-103], clip=[-6,9])
    plot.plot_feature('$project03', space='trajectory', feature='motif', name='GATA1', angles=[58,-103], clip=[-10,10])
    plot.plot_feature('$project03', space='trajectory', feature='motif', name='CEBPB', angles=[58,-103], clip=[-10,10])
    plot.plot_feature('$project03', space='trajectory', feature='motif', name='TCF4', angles=[58,-103], clip=[-5,10])

**NOTE**: Monocle is very sensitive to the input matrix. For example, when we use different versions of scipy (such as 1.0.0 and 1.2.1), the PCA matrices generated by sklearn will be slightly different (difference < 1.0e-7), and even this small difference will produce completely different shapes of trajectories by monocle. Different versions of many other related libraries/packages in python and R environment will also result in different shapes of trajectories. However, APEC can still accurately profile the developmental and differentiation pathway. Here we recommend that users install scipy=1.0.0 to generate the trajectory more similar to the Figure in our paper.


(2) **project02** is an example of running APEC from the user's own fragment count matrix. Users need to copy "filtered_cells.csv" and "filtered_reads.mtx" in **matrix** folder, and "top_filtered_peaks.bed" in **peak** folder to start the clustering. This project contains single cell samples from the forebrain of adult mice, from "Preissl, S. et al. Single-nucleus analysis of accessible chromatin in developing mouse forebrain reveals cell-type-specific transcriptional regulation. Nat Neurosci 21, 432-439 (2018)".

How to run:

    #### cluster cells by APEC algorithm, take ~40 minutes on one CPU-core of computer ####

    from APEC import clustering, plot
    clustering.build_accesson('$project02', ngroup=700)
    clustering.cluster_byAccesson('$project02')
    plot.plot_tsne('$project02', cell_label='cluster')
    plot.correlation('$project02', cell_label='cluster')

    #### cluster cells by chromVAR algorithm, take ~8 hours on 8-core computer ####

    from APEC import clustering, generate
    generate.motif_matrix('$project02', genome_fa='$reference/mm10_chr.fa',
                          background='$reference/tier1_markov1.norc.txt',
                          meme='$reference/JASPAR2018_CORE_vertebrates_redundant_pfms_meme.txt',
                          np=8)
    clustering.cluster_byMotif('$project02', np=8)

    #### plot motifs on tSNE diagram of APEC. It takes ~1 minute for each motif ####

    from APEC import plot
    plot.plot_feature('$project02', space='tsne', feature='motif', name='NEUROD1', clip=[-5,10])
    plot.plot_feature('$project02', space='tsne', feature='motif', name='OLIG2', clip=[-5,8])
    plot.plot_feature('$project02', space='tsne', feature='motif', name='MEF2C', clip=[-5,8])
    plot.plot_feature('$project02', space='tsne', feature='motif', name='MEIS2', clip=[-4,6])
    plot.plot_feature('$project02', space='tsne', feature='motif', name='Dlx2', clip=[-5,11])
    plot.plot_feature('$project02', space='tsne', feature='motif', name='NOTO', clip=[-6,10])
    plot.plot_feature('$project02', space='tsne', feature='motif', name='Sox2', clip=[-22,27])
    plot.plot_feature('$project02', space='tsne', feature='motif', name='ETS1', clip=[-4,8])


(3) **project03** is an example of running APEC from the raw fastq files. Users only need to copy the raw fastq files in **raw_data** folder to start the analysis. This project contains single cell samples of the leukemic stem and blast cells of two different patients (SU070 and SU353), from "Schep, A.N., Wu, B., Buenrostro, J.D. & Greenleaf, W.J. chromVAR: inferring transcription-factor-associated accessibility from single-cell epigenomic data. Nat Methods 14, 975-978 (2017)". Also, users can skip the mapping and alignment, and try this project from the clustering step by only downloading the **matrix** and **peak** folder.

How to run mapping and alignment with bash script:

    #### get fragment count matrix from raw fastq files, take 10~20 hours on 8-core/32GB computer ####

    bash APEC_prepare_steps.sh -r $project01/raw_data -s $project01 -g hg19 -n 10 -l 8 -p 0.05 -f 800

How to run clustering and feature analysis in python enviroment:

    #### cluster cells by APEC algorithm, take <5 minutes on one CPU-core of computer ####

    from APEC import clustering, plot
    clustering.build_accesson('$project01', ngroup=720)
    clustering.cluster_byAccesson('$project01')
    plot.plot_tsne('$project01', cell_label='notes')
    plot.plot_tsne('$project01', cell_label='cluster')

    #### cluster cells by chromVAR algorithm, take ~30 minutes on 8-core computer ####

    from APEC import clustering, generate
    generate.motif_matrix('$project01', genome_fa='$reference/hg19_chr.fa',
                          background='$reference/tier1_markov1.norc.txt',
                          meme='$reference/JASPAR2018_CORE_vertebrates_redundant_pfms_meme.txt',
                          np=8)
    clustering.cluster_byMotif('$project01', np=8)

    #### get differential accessons/genes/motifs for cell cluster 1, take several minutes ####

    from APEC import generate
    generate.gene_score('$project01', genome='hg19')
    generate.differential_feature('$project01', feature='accesson', target='1', vs='all')
    generate.differential_feature('$project01', feature='motif', target='1', vs='all')
    generate.differential_feature('$project01', feature='gene', target='1', vs='all')

    #### plot enrichment of motif RUNX1 on tSNE diagram, take several minutes ####

    from APEC import plot
    plot.plot_feature('$project01', space='tsne', feature='motif', name='RUNX1')

    #### search for potential super enhancer, take several minutes ####

    from APEC import generate
    generate.search_super_enhancer('$project01', super_range=1000000)
