Please visit the following website to download example projects (including project01, project02, and project03):

http://galaxy.ustc.edu.cn:30803/APEC/

**(1)** **project01** contains single cell samples from hematopoietic stem cell differentiation project (including HSC, MPP, CMP, GMP, MEP, LMPP, CLP, pDC, and UNK cells), from "Buenrostro, J.D. et al. Integrated Single-Cell Analysis Maps the Continuous Regulatory Landscape of Human Hematopoietic Differentiation. Cell 173, 1535-1548 e1516 (2018)". Users can run "python script_for_project01.py" to obtain Fig. 1b, 1d, 3c, and 3d of the APEC paper, and this figures will be placed in $project01/figure/ folder.

**NOTE**: In "script_for_project01.py", "project01/" should be the path to the project01 folder; "../reference/" should be the **reference** folder. The reference folder should be filled with the following files downloaded from http://galaxy.ustc.edu.cn:30803/APEC/reference/:

    hg19_RefSeq_genes.gtf, hg19_chr.fa, hg19_chr.fa.fai,
    mm10_RefSeq_genes.gtf, mm10_chr.fa, mm10_chr.fa.fai,
    JASPAR2018_CORE_vertebrates_non-redundant_pfms_meme.txt, tier1_markov1.norc.txt

**NOTE**: Monocle is very sensitive to the input matrix. For example, when we use different versions of scipy (such as 1.0.0 and 1.2.1), the PCA matrices generated by sklearn will be slightly different (difference < 1.0e-7), and even this small difference will produce completely different shapes of trajectories by monocle. Different versions of many other related libraries/packages in python and R environment will also result in different shapes of trajectories.


**(2)** **project02** contains single cell samples from the forebrain of adult mice, from "Preissl, S. et al. Single-nucleus analysis of accessible chromatin in developing mouse forebrain reveals cell-type-specific transcriptional regulation. Nat Neurosci 21, 432-439 (2018)". Users can run "python script_for_project02.py" to obtain Fig. 2a, 2b, and 2e of the APEC paper, and this figures will be placed in $project02/figure/ folder.

If users want to re-plot Fig. 2c, 2d, 2f, and 2g of the APEC paper, they can run "script_python_for_Figure_2c_2d.py" and "script_python_for_Figure_2f_2g.py" after the running of "script_for_project02.py". The output png files "cluster_vs_gene.png" is Fig. 2c, "cluster_pearson_corr.png" is Fig. 2d, "clusters_vs_Excl_types.png" is Fig. 2f, "clusters_vs_Int_types.png" is Fig. 2g.

**NOTE**: In all these scripts, "project02/" should be the path to the project02 folder; "../reference/" should be the **reference** folder.


**(3)** **project03** contains single cell samples of the leukemic stem and blast cells of two different patients (SU070 and SU353), from "Schep, A.N., Wu, B., Buenrostro, J.D. & Greenleaf, W.J. chromVAR: inferring transcription-factor-associated accessibility from single-cell epigenomic data. Nat Methods 14, 975-978 (2017)". Users can run "python script_for_project03.py" to obtain Fig. S3b of the APEC paper.

**NOTE**: In "script_for_project03.py", "project03/" should be the path to the project03 folder.
